compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs

```diff
@@ -161,24 +175,32 @@ fn make_mir_scope<'ll, 'tcx>(
         // Note further that we can't key this hashtable on the span itself,
         // because these spans could have distinct SyntaxContexts. We have
         // to key on exactly what we're giving to LLVM.
-        match discriminators.entry(callsite_span.lo()) {
+        let inlined_at = match discriminators.entry(callsite_span.lo()) {
             Entry::Occupied(mut o) => {
                 *o.get_mut() += 1;
                 unsafe { llvm::LLVMRustDILocationCloneWithBaseDiscriminator(loc, *o.get()) }
-                    .expect("Failed to encode discriminator in DILocation")
             }
             Entry::Vacant(v) => {
                 v.insert(0);
-                loc
+                Some(loc)
+            }
+        };
+        match inlined_at {
+            Some(inlined_at) => {
+                debug_scope.as_mut().unwrap().inlined_at = Some(inlined_at);
+            }
+            None => {
+                // LLVM has a maximum discriminator that it can encode (currently
+                // it uses 12 bits for 4096 possible values. If we exceed that
+                // there is little we can do but drop the debug info and warn.
+                debug_scope = None;
+                tracing::warn!(
+                    "Debug information was lost because LLVM discriminator limits were exceeded"
+                );
```

Comment:
Question: this is a `tracing::warn` that is for rustc developers. But shouldn't this be a user-visible warning?

Reply:
Maybe. It's not particularly actionable to users and we don't warn in all the other cases where we don't emit debug info.

Reply:
> we don't warn in all the other cases where we don't emit debuginfo.

I feel like as a user I would want to have such a warning, but maybe that can get very noisy. Anyway, if we do want to warn when not emitting debuginfo then we should do so consistently, so not for this PR.